ruby/phase-1.md

```diff
@@ -0,0 +1,575 @@
+
+
+## A basic end-to-end example
+
+Perhaps a bit unconventionally (as far as the proposal review process goes), but we'd like to start
+with an end-to-end example of a simple and fully functional "Hello World" workflow. The intention of
+this example is to demonstrate how everything ties together before diving into each separate moving
+part and exploring it in depth.
+
+```ruby
+require 'temporal'
+require 'temporal/worker'
+
+class MyActivity < Temporal::Activity
+  def execute(name)
+    puts "Hello World, #{name}!"
+  end
+end
+
+class MyWorkflow < Temporal::Workflow
+  def execute(name)
+    # This part is outside of the scope of the 1st phase of the proposal
+    MyActivity.execute!(name, options: { task_queue: 'my-queue' })
+  end
+end
+
+config = Temporal::Configuration.new(
+  url: 'localhost:7233'
+)
+
+client = Temporal::Client.new(config)
+client.start_workflow(MyWorkflow, 'Alice', options: {
+  namespace: 'my-namespace',
+  task_queue: 'my-queue'
+})
+
+worker = Temporal::Worker.new(config, 'my-namespace', 'my-queue')
+worker.register_workflow(MyWorkflow)
+worker.register_activity(MyActivity)
+worker.run
+```
+
+
+## Client
+
+The client is the primary point for interacting with workflows and async activities as well as
+managing namespaces. A client is created with a configuration object that specifies options shared
+between a worker and a client.
+
+```ruby
+config = Temporal::Configuration.new(url: 'localhost:7233')
+client = Temporal::Client.new(config)
+client.start_workflow(MyWorkflow, 'arg_1', kwarg_2: 'foo', options: {
+  namespace: 'my-namespace',
+  task_queue: 'my-queue'
+})
+```
+
+Notes:
+
+- Client is intended to be multi-tenant capable (no limit on a single namespace)
+- Configuration will provide defaults where possible (avoiding extensive upfront configuration)
+- Configuration will be used for other more advanced options such as logger, metrics, payload
+  converters, middleware, etc
+
+```ruby
+class Temporal::Client
+  # Creates a new instance of a Client
+  def self.new(config: Temporal::Configuration, **options) -> Temporal::Client
+
+  # Start the workflow and return handle for further interaction
+  # TODO: expand on :options
+  def start_workflow(
+    workflow, # workflow class or a name
+    *args, # any args to be passed as the input to the workflow
+    options: {}, # execution options
+    **kwargs # keyword arguments as inputs are also supported
+  ) -> Temporal::WorkflowHandle
+
+  # Start the workflow and block awaiting it's completion
+  def execute_workflow(
+    workflow, # workflow class or a name
+    *args, # any args to be passed as the input to the workflow
+    options: {}, # execution options
+    **kwargs # keyword arguments as inputs are also supported
+  ) -> Any
+
+  # Generate a handle object from a namespace, workflow_id and a run_id
+  def workflow_handle(namespace, workflow_id, run_id = nil) -> Temporal::WorkflowHandle
```

Comment:
Another thing to consider here is `first_execution_run_id`, if provided to the handle it will guarantee that methods that support it will not affect workflow runs that are not part of the chain that started from first_execution_run_id.

At the moment this can only be used in terminate and cancel requests but we plan to add this safety mechanism to all handle APIs (once it's supported server-side).

Reply:
Just to give more context here:
If I start a workflow and it continues as new, the new run's `first_execution_run_id` will be the original run's ID.
In contrast, if I start the same workflow (with same ID) from a client after the first run completes `first_execution_run_id` will be the "current" run ID.

Reply:
Thank you for explaining this! 🙌 

How do you think it affects the handle? I thought that you want to simply reference a specific `run_id`, which I assume is unique in all cases. So the confusion arises when you skip `run_id`, is that correct? Would you say that in some cases we need to reference an execution based on the `workflow_id` and a `first_execution_run_id `?

Reply:
Yes, sometimes you'd want to reference a workflow by `workflow_id` and `first_execution_run_id`.